spring-kafka/src/main/java/org/springframework/kafka/config/StreamsBuilderFactoryBean.java

```diff
@@ -290,7 +290,15 @@ public synchronized void start() {
 			try {
 				Assert.state(this.streamsConfig != null || this.properties != null,
 						"'streamsConfig' or streams configuration properties must not be null");
-				Topology topology = getObject().build(); // NOSONAR
```

Comment:
We need to keep such a `// NOSONAR` message here. Just because Sonar thinks that `getObject()` may return `null`.

I would change your code to this:
```
Properties topologyProps = this.properties;
if (topologyProps == null) {
    topologyProps = new Properties();
	topologyProps.putAll(this.streamsConfig.originals());
}
Topology topology = getObject().build(); // NOSONAR
```

Much shorter and identical solution.

Although I doubt that `streamsConfig.originals()` returns all the configured properties in the `StreamsConfig`.
Not sure, though, that there is some other way...
Also, please, add your name to the `@author` list and would be great to have some unit-test to be sure that we are covered.

Thanks

Reply:
@artembilan  Thanks for your quick review. I changed the control flow as you suggested, and will think of a unit test that covers the new behavior.

Regarding `streamsConfig.originals()`, the source code of `AbstractConfig` strongly suggests that it returns all configuration values. The simply values are simply not "processed" yet (e.g., qualified class names converted to `Class` objects), but that's the exact raw format `StreamsBuilder#build` expects.

Reply:
I tried adding a unit test, but found it impossible to test the actual effect of passing down `topology.optimization` with the current data `StreamsBuilderFactoryBean` exposes. I'd need to make the created `Topology` accessible to the test, perhaps through a package-private getter.

Reply:
M-m-m. Haven't you looking into a `TopologyTestDriver` from the `kafka-streams-test-utils`: https://kafka.apache.org/11/documentation/streams/developer-guide/testing.html?

Reply:
`TopologyTestDriver` requires a `Topology` object, which I don't have in the test. :) Also, `TopologyTestDriver` operates in a black box fashion and can thus not be used to find out whether topology optimization has been applied (because the point of the optimizations is that they don't change the topology's semantics).

As mentioned before, the only workable solution I could think of is to add a package-private `getTopology()` method to `StreamsBuilderFactoryBean`, and then `describe()` the topology to assert e.g. the absence of a separate changelog topic for a `KTable` source (the easier of the two optimizations to test for).

Reply:
OK. I see what you mean.
So, how about a simple unit test with mocks:
```
StreamsBuilderFactoryBean factoryBean = new StreamsBuilderFactoryBean(config) {
    
        @Override
	protected StreamsBuilder createInstance() {
               return spy(super.createInstance());
        }
  
}
```

and verify in the end that `factoryBean.getObject().build(props)` is called and, of course, we can capture that arg and verify that expected properties are propagated to this call.

Optimized topology or not is not our project responsibility. I just don't care on this level what they do with properties I provide. Our goal to have coverage for our own code. That's all.

Thanks for understanding.

Reply:
Ah, good idea with the anonymous subclass! I‘ll go with that approach.

I agree with topology optimization not being this library‘s responsibility, by the way. I just couldn‘t think of any other observable effect to test on. The spy solves that nicely, though.

Reply:
I now added tests for both types of property sources (`this.properties` and `this.streamsConfig`).